### DISCUSSION TITLE
Simplify optics hierarchy

### DISCUSSION THREAD

**Discussion by serras** [arrow-kt/arrow#2859] [Ideas]
In https://github.com/arrow-kt/arrow/pull/2851 I proposed to align the optics hierarchy in Arrow 2.0 to those in other communities. In particular, get rid of some confusion like `Traversal` being essentially a `Setter` and `Every` being what other call `Traversal`. I think this is a **must have**, otherwise Arrow stays the weird one in the crowd.

![hierarchy optics - page 1](https://user-images.githubusercontent.com/309334/205589282-03681ab6-cb02-45bf-a622-78491120e4ea.png)

However, I think there's room for **further simplification**, based on the way people use Optics. More concretely, I propose to reduce the API to the four types marked in the picture above. The documentation is then simplified by simply talking about "focusing", and having no distinction between "set" and "get":
- `Traversal` focuses in any amount of elements, perfect for working with collections,
- `Optional` is 0 or 1, perfect for focusing on one choice out of a `sealed` hierarchy,
- `Lens` in exactly 1, perfect for focusing on fields,
- `Prism` is an `Optional` which can be "reversed", which is "the other side" of a `sealed` hierarchy.

In addition, I propose to change the derivation of optics for nullable fields. Right now for `amount: Int?` we generate an `Optional`, but I think people find it surprising that it only "sets" when the field was already not null. Instead of this, I propose to generate a `Lens`, and introduce a new combinator,

```kotlin
fun <S, A> Lens<S, A?>.notNull(): Optional<S, A>
```

so "going under a `null`" is explicit in the same way that you need to call `every` to "go under a list".

An additional advantage is that this streamlined hierarchy allows us to reduce the binary size, a common complaint when using Optics. The key problem is that before we generated extension fields for every possible combination,

```kotlin
Iso<S, A>.field: Lens<S, A>
Lens<S, A>.field: Lens<S, A>
// ... 6 more
```

but now we only have 4, so we've effectively halved the size.

I'm happy to work on this, both on the library side and the compiler plug-in side.

**Comment by nomisRev** [comment]
Thanks for the proposal @serras 🙏 

Looks good to me! We should focus on the practical, and what people are using. I like your proposal for `notNull()`, and it would indeed fit nicely into the DSL together with `every`, `filter`, `index` & `at`.
We should probably also re-evaluate the documentation to be more practically oriented. It's perhaps the oldest documentation in Arrow 😅

> but now we only have 4, so we've effectively halved the size.

🥳, I think we can also ditch the extension for `Prism<S, A>.field` since `lens compose prism` is `Optional`, right? So we only need to DSL to be projected over `Lens`, `Optional` and `Traversal`. This would reduce it a bit further.

We should also fix this issue while tackling this, https://github.com/arrow-kt/arrow/issues/2668.

**Comment by serras** [comment]
I'm working on it at https://github.com/arrow-kt/arrow/pull/2873

**Reply by serras** [reply]
This is merged and it's now available in 2.0 snapshots
